---
title: Microtomographic investigation of a large corpus of cichlids
keywords:
- cichlids
- fish morphology
- X-ray micro-tomography
lang: en-US
date-meta: '2023-07-17'
author-meta:
- David Haberthür
- Mikki Law
- Kassandra Ford
- Marcel Häsler
- Ole Seehausen
- Ruslan Hlushchuk
header-includes: |-
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Microtomographic investigation of a large corpus of cichlids" />
  <meta name="citation_title" content="Microtomographic investigation of a large corpus of cichlids" />
  <meta property="og:title" content="Microtomographic investigation of a large corpus of cichlids" />
  <meta property="twitter:title" content="Microtomographic investigation of a large corpus of cichlids" />
  <meta name="dc.date" content="2023-07-17" />
  <meta name="citation_publication_date" content="2023-07-17" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="David Haberthür" />
  <meta name="citation_author_institution" content="Institute of Anatomy, University of Bern, Baltzerstrasse 2, CH-3012 Bern, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0003-3388-9187" />
  <meta name="citation_author" content="Mikki Law" />
  <meta name="citation_author_institution" content="Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland" />
  <meta name="citation_author_institution" content="Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland" />
  <meta name="citation_author" content="Kassandra Ford" />
  <meta name="citation_author_institution" content="Department of Biological Sciences, George Washington University, USA" />
  <meta name="citation_author_institution" content="Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland" />
  <meta name="citation_author_institution" content="Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0002-4413-2633" />
  <meta name="citation_author" content="Marcel Häsler" />
  <meta name="citation_author_institution" content="Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland" />
  <meta name="citation_author_institution" content="Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland" />
  <meta name="citation_author" content="Ole Seehausen" />
  <meta name="citation_author_institution" content="Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland" />
  <meta name="citation_author_institution" content="Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0001-6598-1434" />
  <meta name="citation_author" content="Ruslan Hlushchuk" />
  <meta name="citation_author_institution" content="Institute of Anatomy, University of Bern, Baltzerstrasse 2, CH-3012 Bern, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0002-6722-8996" />
  <link rel="canonical" href="https://habi.github.io/EAWAG-manuscript/" />
  <meta property="og:url" content="https://habi.github.io/EAWAG-manuscript/" />
  <meta property="twitter:url" content="https://habi.github.io/EAWAG-manuscript/" />
  <meta name="citation_fulltext_html_url" content="https://habi.github.io/EAWAG-manuscript/" />
  <meta name="citation_pdf_url" content="https://habi.github.io/EAWAG-manuscript/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://habi.github.io/EAWAG-manuscript/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://habi.github.io/EAWAG-manuscript/v/1f10c12e87d556170c61c2bcd4900dcecb5035ed/" />
  <meta name="manubot_html_url_versioned" content="https://habi.github.io/EAWAG-manuscript/v/1f10c12e87d556170c61c2bcd4900dcecb5035ed/" />
  <meta name="manubot_pdf_url_versioned" content="https://habi.github.io/EAWAG-manuscript/v/1f10c12e87d556170c61c2bcd4900dcecb5035ed/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
bibliography:
- content/manual-references.bib
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
manubot-clear-requests-cache: false
...



_A preprint of this manuscript is available at bioRχiv with the [doi:10.1101/2023.03.30.534917](https://doi.org/10.1101/2023.03.30.534917)._
_We submitted [`v1.0`](https://github.com/habi/EAWAG-manuscript/releases/tag/v1.0) ([online version](https://habi.github.io/EAWAG-manuscript/v/e1e2ef76a476174a4115937d77457037ddec95df/)) of the manuscript to [PLOS ONE](https://journals.plos.org/plosone/)._
_The version you see here [has been updated](https://github.com/habi/EAWAG-manuscript/compare/v1.0...1f10c12e87d556170c61c2bcd4900dcecb5035ed) since submission._


<small><em>
This manuscript
([permalink](https://habi.github.io/EAWAG-manuscript/v/1f10c12e87d556170c61c2bcd4900dcecb5035ed/))
was automatically generated
from [habi/EAWAG-manuscript@1f10c12](https://github.com/habi/EAWAG-manuscript/tree/1f10c12e87d556170c61c2bcd4900dcecb5035ed)
on July 17, 2023.
</em></small>

## Authors



+ **David Haberthür**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon width=16 height=16}
    [0000-0003-3388-9187](https://orcid.org/0000-0003-3388-9187) ·
    ![GitHub icon](images/github.svg){.inline_icon width=16 height=16}
    [habi](https://github.com/habi)
    · ![Mastodon icon](images/mastodon.svg){.inline_icon width=16 height=16}
    [habi](https://mastodon.social/@habi)<br>
  <small>
     Institute of Anatomy, University of Bern, Baltzerstrasse 2, CH-3012 Bern, Switzerland
  </small>

+ **Mikki Law**<br>
    ![GitHub icon](images/github.svg){.inline_icon width=16 height=16}
    [mlaw-migalig](https://github.com/mlaw-migalig)<br>
  <small>
     Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland; Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland
  </small>

+ **Kassandra Ford**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon width=16 height=16}
    [0000-0002-4413-2633](https://orcid.org/0000-0002-4413-2633) ·
    ![GitHub icon](images/github.svg){.inline_icon width=16 height=16}
    [kassthefish](https://github.com/kassthefish)<br>
  <small>
     Department of Biological Sciences, George Washington University, USA; Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland; Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland
  </small>

+ **Marcel Häsler**<br>
    ![GitHub icon](images/github.svg){.inline_icon width=16 height=16}
    [mphaesler](https://github.com/mphaesler)<br>
  <small>
     Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland; Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland
  </small>

+ **Ole Seehausen**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon width=16 height=16}
    [0000-0001-6598-1434](https://orcid.org/0000-0001-6598-1434) ·
    ![GitHub icon](images/github.svg){.inline_icon width=16 height=16}
    [OleSeehausen](https://github.com/OleSeehausen)<br>
  <small>
     Aquatic Ecology and Evolution, Institute of Ecology and Evolution, University of Bern, CH-3012 Bern, Switzerland; Department of Fish Ecology and Evolution, Eawag, Swiss Federal Institute for Aquatic Science and Technology, CH-6047 Kastanienbaum, Switzerland
  </small>

+ **Ruslan Hlushchuk**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon width=16 height=16}
    [0000-0002-6722-8996](https://orcid.org/0000-0002-6722-8996) ·
    ![GitHub icon](images/github.svg){.inline_icon width=16 height=16}
    [RuslanHlushchuk](https://github.com/RuslanHlushchuk)<br>
  <small>
     Institute of Anatomy, University of Bern, Baltzerstrasse 2, CH-3012 Bern, Switzerland
  </small>



## Abstract {.page_break_before}

A large collection of cichlids (N=133) from Lake Victoria in Africa spanning a length range of 6 to 18 cm was nondestructively imaged using micro-computed tomography.
We describe our method to efficiently obtain three-dimensional tomographic datasets of the oral and pharyngeal jaws and the whole skull of these fishes for accurately describing their morphology.
The tomographic data we acquired (9.8 TB of projection images) yielded 1.5 TB of three-dimensional image stacks used for extracting the relevant features of interest.
Herein we present our method and an outlook on analyzing the acquired data; a morphological description of the oral and pharyngeal jaws of the fishes, a three-dimensional geometric morphometrics analysis of landmark features on the fish skulls, and a robust method to automatically extract the otoliths of the fishes from the tomographic data.


## Introduction {.page_break_before}

### History and rationale

Cichlid fish in African lakes are powerful model systems in speciation and adaptive evolutionary radiation research [@doi:10.1038/nrg1316; @doi:10.1098/rspb.2006.3539; @doi:10.1002/ece3.7359].
The functional decoupling of their oral and pharyngeal jaws is hypothesized to be a factor in making cichlids unusually versatile in their feeding and allowing them the ability to adapt to a wide range of environmental factors [@doi:10.2307/2412950].
The hypothesis is that the fusion of the lower pharyngeal jaws makes that jaw system a powerful food processing tool, in turn releasing the oral jaws from functional constraint.
The oral jaws no longer need to process prey and can therefore specialize on prey capture.

Even though the evolutionary diversification of cichlid fish radiations in Lake Victoria in Africa is a well researched issue it remains a complex system in need of further study [@doi:10.1098/rspb.2006.3539; @doi:10.1111/nph.13450].
We aim to better understand the functional anatomy of the skulls and jaws in these fish in order to test the functional decoupling and other hypotheses about what may facilitate exceptionally high rates of morphological evolution.

The collection of cichlid fish available is extremely valuable, hence a nondestructive imaging method is paramount for studying these samples.
Since micro–computed tomography can be regarded as nondestructive method for biological samples, it is the ideal imaging method for investigating the oral and pharyngeal jaws as well as the skull features of the fish presented in this study [@doi:10.1093/iob/obad008] (N=133).

### Micro-computed tomography

X-ray microtomography is a valuable tool to gain insights into the inner structure of highly diverse samples, namely for specimens related to the biomedical sciences [@doi:10.1186/s12915-020-0753-2].
Microtomographic imaging has been employed as a method of choice to nondestructively assess the morphology of different fish species, large and small [@doi:10.1093/iob/obad008].
For a small overview of analyses which are possible with X-ray microtomographic imaging in relation to fish biology and morphology, see prior work of the authors of this manuscript [@doi:10.1371/journal.pone.0228333; @doi:10.1093/iob/obac022; @doi:10.1093/iob/obad008] as well as other studies [@https://osf.io/ecmz4].

Depending on the structures of interest, biomedical samples are often only tomographically scanned after being stained with a contrast agent [@doi:10.26496/bjz.2014.63], most often employing contrast agents containing heavy metals [@10.1089/zeb.2016.1245].
Due to their inherent contrast difference to the surrounding tissue, the structures of interest we touch upon in this manuscript (teeth and bones, i.e. jaws and skull) are well visualized in native samples.
Enhancing contrast differences by staining prior to tomographic imaging was thus not necessary to for the samples presented herein.


## Materials, Methods and Results {.page_break_before}

### Sample procurement and preparation

The fish were kept in 75% ethanol for long-term storage at the Swiss Federal Institute of Aquatic Science and Technology (Eawag).
They were delivered to the Institute of Anatomy for micro-CT imaging sorted into batches of approximately equal length.

### Ethics statement

The study used specimens from an already existing fish collection, built over decades of research led by Ole Seehausen and continuing until today.
All relevant permits for research, the permits for export from the respective country and import to Switzerland were given and the research has followed modern ethical guidelines for scientific research.

### Micro-computed tomographic imaging

All samples were scanned on two of the three available high-resolution micro-CT machines of the Institute of Anatomy of the University of Bern in Switzerland, a SkyScan 1272 and a SkyScan 2214 (both Bruker microCT, Kontich, Belgium).

The fish were sorted into 'bins' based on their physical size.
We used a custom-made sample-holder to scan each of the fish in our machines.
It was 3D-printed on a Form 2 desktop stereolithography printer (Formlabs, Somerville, Massachusetts, USA), the file for printing the holder is available online [@https://github.com/TomoGraphics/Hol3Drs/blob/master/STL/EAWAG.Fish.stl] as part of a library of sample holders for tomographic scanning of biomedical samples [@doi:10.5281/zenodo.2587555].
The original *OpenSCAD* [@https://openscad.org] file [@https://github.com/TomoGraphics/Hol3Drs/blob/master/EAWAG.Fish.scad] is parametrized to effortlessly generate a file for 3D-printing sample holders to accommodate the varying width, height and length classes of the fish.

In total, we acquired 372 tomographic scans of 133 different specimens.
All the scanning parameters are collected in a table in the [Supplementary Materials]; an overview of the parameters is given below.

Since the fish greatly varied in their length (specimen length varied between 6 cm and 18 cm), the voxel sizes of each of the acquired datasets also varies greatly.
We acquired datasets with (isometric) voxel sizes ranging from 3.5 to 50 μm.
<!---
12  103637  rec   3.499972
182  12319  head_50um_rec  49.998527
--->
Due to the geometric magnification employed by the Bruker SkyScan microCT machines, the voxel size of the resulting dataset is depending on the position of the specimen in the machine.
Natually, the field of view of the resulting dataset also depends on the chosen voxel size.
The chosen voxel size is thus scan-specific.
It was always chosen in a way that the region of interest to be images fit into the available lateral field of view of the microCT machine at the smallest necessary voxel size to resolve the structures of interest.
Since the region of interest (either oral or pharyngeal jaws or the complete skulls of the fish) often did not fit into a single field of view *along* the anteroposterior axis of the fish, we often performed several tomographic scans along the rotation axis of the specimen in the machine (since the specimen were all scanned rotating around the anteroposterior axis).
The projection images from those so-called stacked scans are subsequently automatically merged to one stack of reconstructions images by the reconstruction software.

Depending on the size of the specimen we set the X-ray source voltage to 50--80 kV and---depending on the voltage---to a current between 107 and 200 μA.
The different voltages and currents were chosen according to the Bruker guidelines for ideal image acquisition.
The X-ray spectrum was filtered either by an aluminum filter of varying thickness (either 0.25, 0.5 or 1 mm for increasing specimen size) before digitization to projection images or recorded in an unfiltered way (for smaller specimen).
Projection images spanning either 180° or 360° of sample rotation were acquired in angular steps of 0.1°, 0.15° or 0.2°, also depending on size of the fish.
In total we recorded 9.8 TB of projection images (`TIFF` and `*.iif` files, where the `*.iif` files are for the so-called alignment scans).

All the recorded projection images were subsequently reconstructed into three-dimensional stacks of axial `PNG` images spanning the desired regions of interest of each specimen.
All the specimens were scanned with their mouths facing downward in the sample holder and rotating along their long axis.
We thus manually aligned each of the reconstructed datasets so that the lateral axis of the fish was horizontal in relation to the x and y direction of the reconstructed slices.
We reconstructed the projection images with NRecon (Version 1.7.4.6, Bruker microCT, Kontich, Belgium) with varying ring artifact and beam hardening correction values, depending on each fish (again, all relevant values are listed in the [Supplementary Materials]).
In total, this resulted in 1.5 TB of reconstruction images (a bit more than one million `*rec*.png` files).
On average, each of the scans we performed is made up of about 2700 reconstruction images.
<!---
print('We have a total of %s reconstructions on %s' % (Data['NumberOfReconstructions'].sum(), Root))

We have a total of 1005778 reconstructions on /home/habi/research-storage-iee/microCT

print('This is about %s reconstructions per scan (%s scans, %s fishes)' % (round(Data['NumberOfReconstructions'].sum() / len(Data[Data['NumberOfReconstructions'] > 0])), len(Data[Data['NumberOfReconstructions'] > 0]), len(Data.Fish.unique()) -6 ))

This is about 2704 reconstructions per scan (372 scans, 133 fishes)
--->

A small script [@https://github.com/habi/EAWAG/blob/master/rsync-fishes.sh] was used to generate redundant (archival) copies of the raw projection images and copy all the files to a shared network drive on the 'Research Storage' infrastructure of the University of Bern, enabling collaboration on the data by all authorized persons at the same time.
This automated archival and copying process made it possible to delete the raw projection images on the production system shortly after acquisition, freeing resources and easing data handling.
A subset of the data, namely the log files and reconstructed image stacks were kept on the production system for working with it (as described in [Preparation for analysis] below).

### Data analysis

We wrote a set of *Jupyter* [@https://eprints.soton.ac.uk/403913] notebooks with *Python* code to work with the acquired images and to extract the desired data.
The notebooks were written at the start of the project, to be able to process new scans as soon as they were reconstructed.
Re-runs of the notebook added newly scanned and reconstructed data to the analysis, facilitating an efficient quality check of the scans and batched processing of the data.

All notebooks written for this work are available online [@doi:10.5281/zenodo.6798632] and are extensively commented.
The modularized and interactive format chosen for these notebooks makes them easy to adapt for any other study dealing with tomographic data.
The first author of this study is happy to help other scientists to adapt the notebooks to their use-case.
They notebooks are split into different task performed for this study and are specifically mentioned in the respective sections below.

#### Preparation for analysis

The [main *Jupyter* notebook](https://github.com/habi/EAWAG/blob/master/DisplayFishes.ipynb) for this manuscript dealt with reading all log and image files and preparing overview images of each scan for quality checking and further analysis.

At first, *all* log files of *all* the scan present in the processed folder were read into a *pandas* [@doi:10.5281/zenodo.7093122; @doi:10.25080/Majora-92bf1922-00a] dataframe.
This already enabled us to extract the specimen name and scan.
Since we performed multiple scans for each specimen, i.e. a low resolution scan with large field of view for the whole head and one or two scans in high resolution focusing on the region of the oral and pharyngeal jaws it was necessary to be able to efficiently find the scan to be looked at in detail.
From the log files we extracted the scanning and reconstruction parameters of each performed scan for double-checking them and excluding (or necessarily correcting) any operator-error.
All relevant values for each scan were also saved into the aforementioned ataframe.
This allows for an easy collation of all the relevant values into a table (as shown in the [Supplementary Materials]) at the end of each run of the notebook.

After displaying several parameters of both data acquisition and reconstruction for ruling out any operator error we used *Dask* [@dask] to efficiently access the tomographic data on disk (in the end amounting to a total of nearly a million single images).
On average, each of the tomographic datasets contains around 2700 slices, so the total size of the acquired data (1.5 TB) exceeds the RAM size available on an average high-end workstation.
The use of *Dask* (and more specifically *dask-image* [@https://image.dask.org/]) facilitated on-demand loading of the needed data directly from disk for each of the specimen to be analyzed in each step of the notebooks.

At first, we extracted the central view of each of the three axial directions of the datasets (i.e. 'anteroposterior', 'lateral' and 'dorsoventral' view) and either saved those to disk or loaded them from disk if they were already generated in prior runs of the notebook.
The notebook then also generated the maximum intensity projection (MIP) for each of the anatomical planes and either saved them to disk or loaded them from prior runs.

At the end of the notebook we performed a final sanity check on the MIP images.
In this check we examined the mapping of the gray values of the raw projection images to gray values in the reconstructions, i.e. checked that no overexposed pixels were present.
This is an efficient way for double-checking the gray value mapping, since the MIP images have already been generated in prior steps of the notebook and contain the highest gray values present in all the reconstructed images for each scan.

### Image processing

#### Extraction of oral and pharyngeal jaws, visualization of tomographic data

To extract the oral jaw (OJ) and pharyngeal jaw (PJ) of the fish, we used [*3D Slicer*](https://www.slicer.org) (Version 4.11.20210226) [@doi:10.1016/j.mri.2012.05.001] extended with the *SlicerMorph* tools [@doi:10.1111/2041-210X.13669] which aim to help biologists work with 3D specimen data.
The reconstructed image stacks were loaded into [*ImageStacks*](https://www.slicer.org/wiki/Documentation/Labs/ImageStacks), depending on their size we reduced the image resolution (i.e. downscaled the images) for this first step.
The three-dimensional volume was rendered via [*VTK GPU Ray Casting*](https://slicer.readthedocs.io/en/latest/user_guide/modules/volumerendering.html).
A custom-made volume property was used as an input to view the scans.
Using toggles in the volume rendering, we defined regions of interest (ROIs) for both the OJs and PJs in each specimen.
These ROIs were then extracted in their native resolution from the original dataset for further processing.
Using the gray value thresholding function in *3D Slicer*'s [*Segment Editor*](https://slicer.readthedocs.io/en/latest/user_guide/modules/segmenteditor.html), the teeth in both the oral and pharyngeal jaws were extracted.
We used the *Scissor* and *Island* tools of the *Segment Editor* to isolate single regions.

Processed regions of interest were exported as `Nrrd` [@https://teem.sourceforge.net/nrrd/format.html] files.
The three-dimensional visualizations of all regions of interest for each specimen were compiled into overview images (see Figure @fig:104016 for an example from the compilation document).
In total, we compiled such overview images for 133 specimens with full head morphology, oral jaw and lower pharyngeal jaw profiles.

![Overview of data from sample 104016, *Enterochromis I cinctus*, from 'Station E' in the transect of Mwanza Bay at the southern edge of Lake Victoria.
  A map showing the detailed location of 'Station E' is shown in Fig. 1 of [@doi:10.1007/s10750-016-2925-1].
  Panel A: Three-dimensional visualization of the head scan.
  Panel B: Three-dimensional visualization of the oral jaw scan.
  Panel C: Photograph of the specimen.
  Panels D and E: Three-dimensional visualization of the pharyngeal jaw, dorsoventral and lateral view, respectively.](images/104016.png){#fig:104016}

<!---
- 104016 length is given as '<7 cm' in XLS sheet.
- PJ triangle measured on MIP of 104016_PJ.seg.nrrd, giving 1067.208, 1024.018, 1040.663, i.e. 1045 px. Scanned at 5 um, this gives 5.219815 mm as length in Fiji.
- Cusp distance measured on MIP of 104016_OJcrop.nrrd, giving 41.485, 39.850, 43.932, 48.600, i.e. 45 px. Scanned at 5 um, this gives 217 um.
--->

#### Principal components analysis of skull landmarks

Current studies are using 3D geometric morphometrics to compare the morphological shape of these scanned cichlids using statistical analysis.
We used a homologous landmark scheme across one-half of the skull for higher density of shape information [@doi:10.1643/i2021016; @doi:10.1093/iob/obac022], and landmarks were placed on each specimen using *3D Slicer*.
To examine differences in shape across the species sampled, we performed a Generalized Procrustes Superimposition on the landmark data to remove the effects of location, size, and rotation from the analysis using the *geomorph* package in *R* (Version 4.2.1) with *RStudio* (Version 2022.07.2+576) [@doi:10.1111/2041-210X.13029; @geomorph; @rrpp; @r ;@rstudio].
This process brings all specimens to a common origin, scales the landmarks to a unit centroid size, and rotates specimens to reduce distances between landmarks.
A principal components analysis was then performed in *geomorph* on the superimposed landmark data to visualize the major axes of shape change across sampled species.
We then used phylogenetic information to identify instances of repeated evolution of trophic adaptations in these cichlids.

#### Automatic extraction of otoliths

Otoliths are structures made up of mostly calcium carbonate located in the head of fishes.
Due to their composition, they are easily distinguished in the X-ray images we acquired.
We devised an image processing method to automatically and robustly detect the location of the otoliths in the tomographic scans of the skulls of the cichlids.
Based on this detected location we were then able to efficiently extract a cropped region of interest from the original data which includes only the region of the otoliths.
The whole method is implemented in its own [*Jupyter* notebook](https://github.com/habi/EAWAG/blob/master/ExtractOtoliths.ipynb) (part of the aforementioned analysis repository [@doi:10.5281/zenodo.6798632]).

Since we took great care to scan the fish parallel to their anteroposterior direction and reconstructed the tomographic datasets parallel to the lateral and dorsoventral direction of the fish we could use this 'preparation' for automatically extracting the otoliths the tomographic datasets of the whole fish heads.
By extracting both the peaks and the peak widths of the gray values along both the horizontal and vertical direction of the MIP (generated above) we robustly detected the position of the otoliths in the datasets.
The robust detection is supported by suppressing a small, configurable part of each region, i.e. the front and back, top and bottom or the flanks.

![Visualization of automatic otolith extraction.
  The top row shows the found location of the otolith center in each of the three anatomical directions.
  The bottom row shows a maximum intensity projection of each of the three anatomical directions of the extracted otolith region.](images/104016.head.rec.otolither.position.png){#fig:otolithextraction}

Figure @fig:otolithextraction shows the visualization of the process.
The colored horizontal and vertical bars in each of the directional MIPs denote the found peak location of the two appropriate values.
The white bars show the mean of the to detected positions, which was used for extracting the otoliths from the original datasets.
Making use of the *Dask* library facilitated efficient access to all the data on disk and writing out small, cropped copies of the datasets around the otolith positions.

By detecting the largest components in the cropped copies of the datasets we can easily extract and visualize the otoliths in 3D, as shown in Figure @fig:otolith3d.
The extracted otoliths are thus prepared for further analysis and visualization.
The simple three-dimensional visualization is integrated in the aforementioned *Jupyter* notebook through an integrated visualization library [@https://github.com/K3D-tools/K3D-jupyter] and is also shown in the [Supplementary Materials].

![Three-dimensional view of extracted otoliths of specimen 104016.
  The specimen was scanned with an isotropic voxel size of 13 μm.
  The extracted otolith has a size of approximately 250 x 350 x 150 pixels.
  The axes labels correspond to mm.
  A dynamic view of the visualization is available in the [Supplementary Materials].](images/104016.head.rec.otolith.region.3D.png){#fig:otolith3d}

The notebook for extracting the otoliths can be run in your browser without installing any additional software (via *Binder* [@doi:10.25080/majora-4af1f417-011]).
To do this, one starts the notebook by [clicking a single button](https://mybinder.org/v2/gh/habi/eawag/HEAD?labpath=ExtractOtoliths.ipynb) in the [README file](https://github.com/habi/EAWAG/blob/master/README.md) of the [project repository](https://github.com/habi/EAWAG) [@doi:10.5281/zenodo.6798632].
This starts a computing environment in the cloud, downloads the tomographic data we acquired of *one* specimen, and performs both the otolith extraction and visualization in your browser.


## Discussion {.page_break_before}

We acquired high resolution tomographic datasets of a large collection of cichlids (N=133), several tomographic scans were performed for each specimen.
The 372 acquired datasets were imaged over a wide-spanning range of voxel size (3.5--50 μm) permitting both the analysis of finest details we wanted to resolve (i.e. containing only the oral and pharyngeal jaws) and having datasets (N=104) containing the whole head for principle components analysis and extraction of the otoliths.
In our microCT imaging setup the resulting voxel size is directly correlated with the sample diameter, i.e. smaller specimen were scanned with the smallest voxel size, in the range of several μm.
The skulls of larger specimen were scanned with the largest voxel sizes, in the range of several tens of μm.
The chosen voxel size is thus specimen- or scan-specific and was always selected such that the desired region of interest fits into the available lateral field of view of the microCT machine at the minimally necessary voxel size to resolve the structures of interest.

### Imaging and preparation for analysis

The whole study we presented here spanned a *long* time frame.
It was thus paramount to run the imaging process and the preparation of the tomographic datasets in a batched mode.
The *Jupyter* notebook written to prepare the datasets for quality control and analysis were facilitating a short turnaround time for feedback on single scans and such a batched analysis.

### Otolith extraction

The method to extract the otoliths from the tomographic dataset works robustly for all of the different fish sizes and shapes.
The extraction is robust because it is based on a combination of distinct details of the gray value curve over the different anatomical directions.
The details of the otolith extraction method have been extensively tuned and run in a fully automatic way.
This allows a highly reproducible and unbiased extraction of the otoliths from the tomographic datasets.
This is even the case for one fish which was scanned with still a hook in his mouth, where the otoliths were nonetheless extracted automatically.

Data on such automatically extracted otoliths, like volume and geometric information like eccentricity and moments of inertia is biologically interesting as the otoliths grow with the age of the fish.
One could help estimate the age of wild fishes using a calibration based on the otolith measurements of a fish of known age.
It is worth nothing that estimation of age of tropical fishes is not as simple as for fishes from temperate regions where one can distinguish summer and winter layers within the otolith.

To our best knowledge, no fully automated method for non-destructive otolith extraction from high-resolution tomographic data has been published up to now.
Vasconcelos-Filho et al. [doi:10.3389/fmars.2019.00728] show a potentially objective method to count growth rings in otoliths of six fish species, but performed their study on extracted otoliths "embedded in a styrofoam cube", while our method non-destructively extracts the otoliths from the tomographic data.
Gu et al. [@doi:10.1016/j.micron.2022.103342] also provide detailed insight into otolith and combine microtomographic imaging (with unknown voxel size) with scanning electron microscopy, but also mention that the "otoliths were removed [from *Bahaba taipingensis*], cleaned and kept dry for morphologic observation".

### Outlook

The acquired tomographic datasets are the basis for multiple additional analyses of fish morphology.

The presented method offers an insight and algorithm on how to perform tomographic scans, preview and analyze micro-computer tomographic datasets of a large collection of fish.
The workflow is relying only on free and open-source software and can thus be used and verified independently by any interested reader.
All the *Jupyter* notebook described herein is also freely available online [@doi:10.5281/zenodo.6798632].


## Author Contributions {.page_break_before}

[Contributor Roles Taxonomy](https://credit.niso.org/), as defined by the [National Information Standards Organization](https://credit.niso.org/contributor-roles-defined/).

|Author|Contributions|
|---|---|
|David Haberthür|Conceptualization, Data curation, Formal Analysis, Investigation, Methodology, Project administration, Software, Visualization, Writing – original draft, Writing – review & editing|
|Mikki Law|Data curation, Investigation, Methodology, Project administration, Visualization, Writing – review & editing|
|Kassandra Ford|Investigation, Methodology, Visualization, Writing – review & editing|
|Marcel Häsler|Investigation, Project administration, Resources, Writing – review & editing|
|Ole Seehausen|Conceptualization, Investigation, Resources, Supervision, Writing – review & editing|
|Ruslan Hlushchuk|Conceptualization, Resources, Supervision, Writing – review & editing|

## Acknowledgments

We thank Salome Mwaiko for taking care of the fish collection at Eawag and Mark Charran for helping to find suitable specimens to represent each species.
We are grateful to the [Microscopy Imaging Center](https//www.mic.unibe.ch/) of the University of Bern for the infrastructural support.
We thank the `manubot` project [@doi:10.1371/journal.pcbi.1007128] for helping us write this manuscript collaboratively.


## Supplementary Materials

### Parameters of tomographic scans of all the fishes

The CSV file [ScanningDetails.csv](https://github.com/habi/EAWAG-manuscript/blob/main/content/data/ScanningDetails.csv) gives a tabular overview of all the (relevant) parameters of all the scans we performed.
This file was generated with the [data processing notebook](https://github.com/habi/EAWAG/blob/master/DataWrangling.ipynb) and contains the data which is read from *all* the log files of *all* the scans we performed.
A copy of each log file is available in a [folder in the data processing repository](https://github.com/habi/EAWAG/tree/master/logfiles).

### Three-dimensional view of *one* of the extracted otoliths

The three-dimensional view of sample 104016 was generated in [the otolith extraction notebook](https://github.com/habi/EAWAG/blob/master/ExtractOtoliths.ipynb) and saved as a [self-contained `HTML` file](https://github.com/habi/EAWAG-manuscript/blob/main/content/data/104016.head.rec.otolith.region.3D.html) with [*K3D-jupyter*](https://github.com/K3D-tools/K3D-jupyter).
A copy of this `HTML` file can be viewed and interacted with through the [GitHub HTML preview](https://htmlpreview.github.io/?https://github.com/habi/EAWAG-manuscript/blob/main/content/data/104016.head.rec.otolith.region.3D.html).


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>
